### PR TITLE
CL/BASIC: use ALL available TLs by default

### PR DESCRIPTION
**Pull request by vspetrov**
## What
All CLs (even though we have only 1 at the moment  - basic) have the parameter UCC_CL_BASIC_TLS which defines the TLS the given CL requires. It can be ALL which is set to default for BASIC. CL/BASIC will try to use ALL tls specified with this variable (can be controlled). The logic of tl context/team/score creation in CL/BASIC is unified. 

## Why ?
After that change any 3rd party TL plugin (.so) can be just distributed separately and will be enabled via CL basic OOB according to default scoring settings.

### CODE REVIEW

File: src/components/cl/ucc_cl.c [hunk 1]
```diff
@@ -12,6 +12,9 @@ ucc_config_field_t ucc_cl_lib_config_table[] = {
     {"", "", NULL, ucc_offsetof(ucc_cl_lib_config_t, super),
      UCC_CONFIG_TYPE_TABLE(ucc_base_config_table)},
 
+    {"TLS", "all", NULL, ucc_offsetof(ucc_cl_lib_config_t, tls),
```
Comment (bureddy) [inline src/components/cl/ucc_cl.c:15]: is there a way for users to know possible TLS values? can we have help string instead of NULL?

Comment (vspetrov) [reply src/components/cl/ucc_cl.c]: Added the doc string and available TLS. Plz see how it is done: first i set doc string in config field to NULL. Then on the first call to cl_lib_parser i construct the tls list string (at that moment the TL framework is loaded and i can build the list) and then i reset the "doc" pointer to the constructed message with tls list.
The result ucc_info output:
```
#
# List of TLs used by a given CL component.
# Allowed values: either "all" or comma-separated list of: nccl,ucp
#
# syntax:    comma-separated list of: string
# inherits:  UCC_TLS
#
UCC_CL_BASIC_TLS=all

```

**Review by bureddy — APPROVED** [review]

**Comment by artemry-nv** [comment]
One test failed in Jenkins CI:
```[2021-05-13T20:16:46.132Z] [1620937005.897271] [swx-clx01:167  :0]     ucc_context.c:165  UCC  ERROR failed to modify CL "basic" configuration, name TEST_PARAM, value 12345
[2021-05-13T20:16:46.132Z] /opt/nvidia/torch-ucc/src/ucc/test/gtest/core/test_context_config.cc:54: Failure
[2021-05-13T20:16:46.132Z] Value of: ucc_context_config_modify(ctx_config, "basic", "TEST_PARAM", "12345")
[2021-05-13T20:16:46.132Z]   Actual: -6
[2021-05-13T20:16:46.132Z] Expected: UCC_OK
[2021-05-13T20:16:46.132Z] Which is: 0
[2021-05-13T20:16:46.132Z] /opt/nvidia/torch-ucc/src/ucc/test/gtest/core/test_context_config.cc:58: Failure
[2021-05-13T20:16:46.132Z] Expected: (std::string::npos) != (output.find("UCC_CL_BASIC_TEST_PARAM=12345")), actual: 18446744073709551615 vs 18446744073709551615
[2021-05-13T20:16:46.132Z] [  FAILED  ] test_context_config.modify, where TypeParam =  and GetParam() =  (30 ms)```

**Comment by vspetrov** [comment]
@Sergei-Lebedev @manjugv plz review

File: src/components/cl/basic/cl_basic_context.c [hunk 2]
```diff
@@ -11,37 +11,48 @@ UCC_CLASS_INIT_FUNC(ucc_cl_basic_context_t,
...
+    for (i = 0; i < tls->count; i++) {
+        status = ucc_tl_context_get(params->context, tls->names[i],
+                                    &self->tl_ctxs[self->n_tl_ctxs]);
+        if (UCC_OK != status) {
+            cl_info(cl_config->cl_lib,
+                    "TL %s context is not available, skipping", tls->names[i]);
+        } else {
+            self->n_tl_ctxs++;
+        }
+    }
```
Comment (Sergei-Lebedev) [inline src/components/cl/basic/cl_basic_context.c:43]: check that self->n_tl_ctxs is not 0 after loop

**Review by Sergei-Lebedev — APPROVED** [review]

**Comment by vspetrov** [comment]
bot:retest

**Pull request merged**